dsp/modules/google_vertex.py

```diff
@@ -0,0 +1,164 @@
+    return True
+
+
+BLOCK_ONLY_HIGH = {
+    HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
+    HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
+}
+
+class GoogleVertex(LM):
+    """Wrapper around Google Cloud's Vertex API.
+
+    Supports all text+ models accessible through
+        https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models
+    """
+
+    def __init__(
+            self,
+            project_id: str,
+            region: str = "us-central1",
+            model: str = "gemini-1.0-pro",
+            credentials = None,
+            safety_settings: Optional[Iterable] = BLOCK_ONLY_HIGH,
+            **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        project_id : str
+            Which GCP project to use?
+        model : str
+            Which pre-trained model from Google to use?
+            Choices can be found at https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models
+        **kwargs: dict
+            Additional arguments to pass to the API provider.
+        """
+        super().__init__(model)
+
+        vertexai.init(project=project_id, location=region, credentials=credentials)
+
+        # Google API uses "candidate_count" instead of "n" or "num_generations"
+        # Gemini only supports 1 generation at a time. Raises an error if candidate_count > 1.
+        # Bison supports up to 4.
+        num_generations = kwargs.pop("n", kwargs.pop("num_generations", 1))
+
+        max_output_tokens = kwargs.pop("max_output_tokens", 2048)
+
+        if model.startswith("gemini"):
+            num_generations = 1
+        else:
+            num_generations = 4 # Bison supports upto 4 candidates.
+
+
+        self.provider = "google_vertex"
+
+        kwargs = {
+            "candidate_count": num_generations,
+            "temperature": 0.0 if "temperature" not in kwargs else kwargs["temperature"],
+            "max_output_tokens": 2048,
+            "top_p": 1,
+            "top_k": 1,
+            **kwargs,
+        }
+
+        self.config = GenerationConfig(**kwargs)
+
+        self.llm = GenerativeModel(
+            model_name=model,
+            generation_config = self.config,
+            safety_settings = safety_settings
+            )
+        
+        self.kwargs = {
+            "n": num_generations,
+            **kwargs,
+        }
+
+        self.history: list[dict[str, Any]] = []
+
+
+    def basic_request(self, prompt: str, **kwargs):
+        raw_kwargs = kwargs
+        kwargs = {
+            **self.kwargs,
+            **kwargs,
+        }
+
+        # Google disallows "n" arguments
+        n = kwargs.pop("n", None)
+        if n is not None and n > 1 and kwargs['temperature'] == 0.0:
+            kwargs['temperature'] = 0.7
+
```

Comment:
On retries, `dsp` adds `max_tokens` which will error out on the next line because of VertexAI's use of `proto`. 
So, my suggestion is to add `_ = kwargs.pop("max_tokens")` here